### PR TITLE
mirror: Replicate versions when enabled

### PR DESCRIPTION
**Pull request by vadmeste**
Spec not approved yet.

Still in testing phase

### CODE REVIEW

File: cmd/client-errors.go [hunk 1]
```diff
@@ -214,3 +214,27 @@ type SameFile struct {
 func (e SameFile) Error() string {
 	return fmt.Sprintf("'%s' and '%s' are the same file", e.Source, e.Destination)
 }
+
+// RetentionAlreadyExists - is thrown when we don't want to overwrite existing retention config
+type RetentionAlreadyExists struct {
```
Comment (klauspost) [inline cmd/client-errors.go:219]:
```suggestion
type ErrRetentionAlreadyExists struct {
```

File: cmd/client-errors.go [hunk 2]
```diff
@@ -214,3 +214,27 @@ type SameFile struct {
...
+}
+
+func (e RetentionAlreadyExists) Error() string {
+	return "Retention already exists in the target location."
+}
+
+// LegalholdAlreadyExists is thrown when we don't want to overwrite existing legalhold config
+type LegalholdAlreadyExists struct {
```
Comment (klauspost) [inline cmd/client-errors.go:227]:
```suggestion
type ErrLegalholdAlreadyExists struct {
```

File: cmd/client-errors.go [hunk 3]
```diff
@@ -214,3 +214,27 @@ type SameFile struct {
...
+}
+
+func (e LegalholdAlreadyExists) Error() string {
+	return "Legalhold already exists in the target location."
+}
+
+// TagsAlreadyExists is thrown when we don't want to overwrite existing tags
+type TagsAlreadyExists struct {
```
Comment (klauspost) [inline cmd/client-errors.go:235]:
```suggestion
type ErrTagsAlreadyExists struct {
```

File: cmd/client.go [hunk 4]
```diff
@@ -82,6 +82,22 @@ type CopyOptions struct {
 	isPreserve       bool
 }
 
+// PutOptions holds options for object putting operation
+type PutOptions struct {
```
Comment (klauspost) [inline cmd/client.go:86]: Exporting this but none of the fields does not really make it usable.

File: cmd/client.go [hunk 5]
```diff
@@ -82,6 +82,22 @@ type CopyOptions struct {
...
+	metadata                          map[string]string
+	sse                               encrypt.ServerSide
+	md5, disableMultipart, isPreserve bool
+	versionID, etag                   string
+	modTime                           time.Time
+	isDeleteMarker                    bool
+}
+
+// RemoveOptions holds options for removing object operation
+type RemoveOptions struct {
```
Comment (klauspost) [inline cmd/client.go:96]: Same...

File: cmd/common-methods.go [hunk 6]
```diff
@@ -276,50 +277,82 @@ func getSourceStream(ctx context.Context, alias, urlStr, versionID string, fetch
...
-	if ok {
-		lockMode = minio.RetentionMode(lockModeStr)
-		delete(metadata, AmzObjectLockMode)
+	tagsMaps, err := clnt.GetTags(ctx, versionID)
+	if err != nil {
+		return "", err.Trace(alias, urlStr)
 	}
+	var tags string
+	for k, v := range tagsMaps {
+		tags += fmt.Sprintf("%s=%s&", k, v)
```
Comment (klauspost) [inline cmd/common-methods.go:292]: No escaping needed? What if either contains an `&` or `=`?

File: cmd/difference.go [hunk 7]
```diff
@@ -160,18 +160,49 @@ func metadataEqual(m1, m2 map[string]string) bool {
...
+	return difference(ctx, sourceClnt, targetClnt, sourceURL, targetURL, false, false, false, true, DirFirst)
 }
 
-func differenceInternal(ctx context.Context, sourceClnt, targetClnt Client, sourceURL, targetURL string, isMetadata bool, isRecursive, returnSimilar bool, dirOpt DirOpt, diffCh chan<- diffMessage) *probe.Error {
+type diffEntry struct {
+	path    string
+	modTime time.Time
+}
+
+func cmpDiffEntry(e1, e2 diffEntry) int {
```
Comment (klauspost) [inline cmd/difference.go:176]: Doc

File: cmd/client-errors.go (hunk 1 above)
Comment (vadmeste) [reply cmd/client-errors.go:219]: @klauspost this doesn't follow the existing style, maybe we can do that in another PR

File: cmd/client.go (hunk 4 above)
Comment (vadmeste) [reply cmd/client.go:86]: > Exporting this but none of the fields does not really make it usable.

Still this follows the existing style. I can send other PRs once this merged (this PR is not that easy to review in the first place)

File: cmd/client-errors.go (hunk 1 above)
Comment (klauspost) [reply cmd/client-errors.go:219]: It follows all other styles. No need to introduce exported types that will be renamed.

File: cmd/common-methods.go (hunk 6 above)
Comment (vadmeste) [reply cmd/common-methods.go:292]: > No escaping needed? What if either contains an & or =?

Done

File: cmd/difference.go (hunk 7 above)
Comment (vadmeste) [reply cmd/difference.go:176]: > Doc

Done

**Pull request closed without merging**